### PR TITLE
Revamp snapshot tests

### PR DESCRIPTION
**Pull request by jridgewell**
This fixes quite a few warts with out snapshot tests:

1. Implements good diffing via [similar](https://docs.rs/similar/latest/similar) crate. 
   - It'll look just like git diffs now.
2. Removes the HMR chunk item source maps
   - So, you'll only see 1 source map per emitted chunk file, not 1 per item in the chunk
3. Makes source map files have a consistent fingerprint path
   - Every time the associated JS file contents changed, the `.map` file is renamed, making git show it as a new/deleted file instead of a changed one.
4. Extracts snapshot tests into a separate crate
   - Due to a quirk in rust `build.rs`, we had to rebuild the `turbopack` crate whenever a test file changed
   - This slowed down the next test run, because everything that depends on `turbopack` (read, everything) needed to recompile
   - This slowed down the next build, for the same reason
   - Now, only the separate `turbopack-tests` crate rebuilds, and everything else stays constant

### CODE REVIEW

File: crates/turbopack-tests/README.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# turbopack-tests
+
+An extracted create to perform snapshot tests on turbopack.
```
Comment (alexkirsz) [inline crates/turbopack-tests/README.md:3]:
```suggestion
An extracted crate to perform snapshot tests on turbopack.
```

File: crates/turbopack-tests/README.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+It's possible to only run the snapshot tests using [nextest][]'s filter
+expressions:
+
+```bash
+cargo nextest run -E 'test(snapshot)'
+```
+
+The filter supports any substring, and only test names which contain
+that substring will run. Eg,
```
Comment (alexkirsz) [inline crates/turbopack-tests/README.md:15]:
```suggestion
that substring will run.
```

File: crates/turbopack-tests/README.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+## Updating Snapshot
+
+If you've made a change that requires many snapshot updates, you can
+autmaotcially update all outputs using the `UPDATE` command line env:
```
Comment (alexkirsz) [inline crates/turbopack-tests/README.md:20]:
```suggestion
automatically update all outputs using the `UPDATE` command line env:
```

File: crates/turbopack-tests/tests/snapshot.rs [hunk 4]
```diff
@@ -244,12 +242,39 @@ fn remove_file(root: &str, path: &str) -> Result<()> {
...
+
+    if EcmascriptChunkSourceMapAssetVc::resolve_from(asset)
+        .await?
+        .is_some()
+    {
+        // .map files have a hash like foo.js.abc123.map.
+        let mut p = PathBuf::from(&path.await?.path);
+        if p.extension() == Some(OsStr::new("map")) {
+            p.set_extension("");
+            debug_assert_ne!(p.extension(), Some(OsStr::new("js")));
```
Comment (alexkirsz) [inline crates/turbopack-tests/tests/snapshot.rs:261]: In what case is this possible, since we set the extension to "" just before?

**Review by wbinnssmith — APPROVED** [review]

File: crates/turbopack-tests/tests/snapshot.rs [hunk 5]
```diff
@@ -262,51 +287,52 @@ async fn walk_asset(
...
+                println!("contents of {path_str} did not match:");
+            }
+            let expected = expected.unwrap_or_default();
+            let diff = TextDiff::from_lines(&expected, &actual);
+            println!(
+                "{}",
+                diff.unified_diff()
+                    .context_radius(3)
+                    .header("expected", "actual")
+            );
```
Comment (alexkirsz) [inline crates/turbopack-tests/tests/snapshot.rs:334]: Perhaps use `eprintln` instead of `println` in the above?

**Review by alexkirsz — APPROVED** [review]

File: crates/turbopack-tests/tests/snapshot.rs (hunk 4 above)
Comment (jridgewell) [reply crates/turbopack-tests/tests/snapshot.rs:261]: It recomputes afterwards, inheriting the next `.abc123` that still remains. I think it's implemented like an `rfind`.

File: crates/turbopack-tests/tests/snapshot.rs (hunk 5 above)
Comment (jridgewell) [reply crates/turbopack-tests/tests/snapshot.rs:334]: Done.

**Review by ForsakenHarmony — APPROVED** [review]
Any chance we could merge the snapshot output into a single file?

**Comment by jridgewell** [comment]
> Any chance we could merge the snapshot output into a single file?

Does this make it easier? I really dislike jests file format, because there's cruft alongside the snapshot outputs.

**Comment by ForsakenHarmony** [comment]
Would make git diffs a bit nicer if we don't have 10 changed files for every snapshot test all the time

But that doesn't have to be in this PR

**Pull request merged**